### ISSUE TITLE
Add support for SPF records

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3928]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3928|
|Type|enhancement|
|Created|2009-07-24 16:19:14Z|
|Branch|https://github.com/twisted/twisted/tree/spf-3928|

See http://www.ietf.org/rfc/rfc4408.txt

Attachments:

* [Record_SPF.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/137/137d22929d35f70072157667115aebbcd3f81530/dec1a3d33cff9a82b9d16c3aa59bc307c6199031.patch) (7873 bytes) - added by shylent on 2010-10-27 09:57:40Z - 
* [Record_SPF_2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/137/137d22929d35f70072157667115aebbcd3f81530/a100732b19a6424978d8b8c81f9142169791b9a6.patch) (9351 bytes) - added by shylent on 2010-10-28 22:20:22Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3928 3928
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_spf_3928 branches/spf-3928
branch_author__shylent shylent
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1248452354000000 1248452354000000
changetime__1289156310000000 1289156310000000
version__None None
owner__ 
cc__shylent
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="shylent's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|shylent removed owner|
|-|-|

Add support for SPF records.

As per RFC 4408, SPF record format is similar to TXT record format and, as I understand it, it is not up to twisted.names to parse the actual policy declaration. 

I've reused existing TXT record behaviour and added the necessary resolver methods and tests (hopefully).

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_2"></a>|@lvh set owner to shylent|
|-|-|

Hello, and thanks for your patch!

1. Style issue with kwarg notation in interfaces.py: `def lookupSenderPolicy(name, timeout = 10):` should be `timeout=10`, but apparently this will be fixed in [#2336](https://github.com/twisted/twisted/issues/2336) Soon(TM), plus it is consistent with the rest of the source file so not a showstopper
2. I can not find anything about the ttl kwarg or attribute. This appears to be documented on SimpleRecord but Record_TXT (and by extension Record_SPF) does not use it. I think the easiest fix is: document on Record_TXT, point to that in Record_SPF, add attribute on the IRecord interface. The last thing should probably be a separate ticket.
3. I'm not sure about the tests: twisted/names/test/test_dns.py line 1227. I think the records are used wrong which makes the tests silently fail. Here is an example:
```
>>> from twisted.names.dns import Record_TXT
>>> Record_SPF = Record_TXT # handwavy argument
>>> r = Record_SPF(['foo', 'bar'], 10)
>>> r.ttl
>>> # Note how the ttl is None, not 10.
```

Everything else looks okay :-)

**Comment by twisted-trac** [comment]
|<img alt="shylent's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|shylent commented|
|-|-|

Added documentation for ttl instance attribute for Record_TXT and Record_SPF.

Fixed the mentioned tests for both Record_TXT and Record_SPF (wrong `__init__` signature was used).

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)|@lvh removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [30240]) Branching to 'spf-3928'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [30241]) Apply Record_SPF_2.patch

refs [#3928](https://github.com/twisted/twisted/issues/3928)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @exarkun|
|-|-|

A few minor points:

  1. Copyright dates should be updated on modified files
  1. There should be a news fragment explaining the new feature
  1. New tests should follow the current naming/docstring convention (so `testSPF` should be `test_spf`)

These are all trivial, so I made the changes myself.  The only interesting comment I have, I think, is the `IResolver` change.  In general I would shy away from changing a released interface.  Adding methods to implementations is fine, but adding methods to interfaces can suddenly make previously valid, complete implementations suddenly incomplete.  This may break things, so we should probably avoid it.  On top of that, `IResolver` is an ugly interface: it's defined in the wrong place and the way it forces a new method for every record type is cruddy.  We should replace it with something simpler that won't suffer from the same proliferation issue.  I talked about this with glyph and he agreed, so I've reverted the change to the interface.

Aside from that, this looks good, and [buildbot is happy](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/spf-3928), so I'll merge.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [30253]) Merge spf-3928

Author: shylent
Reviewer: lvh, exarkun
Fixes: [#3928](https://github.com/twisted/twisted/issues/3928)

Add basic support for serving and requesting SPF records with Twisted Names.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
